packages/block-editor/src/hooks/border-color.js

```diff
@@ -0,0 +1,197 @@
+
+	if ( useIsBorderColorDisabled( props ) ) {
+		return null;
+	}
+
+	const onChangeColor = ( value ) => {
+		const colorObject = getColorObjectByColorValue( colors, value );
+		const newStyle = {
+			...style,
+			border: {
+				...style?.border,
+				color: colorObject?.slug ? undefined : value,
+			},
+		};
+
+		const newNamedColor = colorObject?.slug ? colorObject.slug : undefined;
+
+		setAttributes( {
+			style: cleanEmptyObject( newStyle ),
+			borderColor: newNamedColor,
+		} );
+	};
+
+	return (
+		<ColorGradientControl
+			label={ __( 'Border color' ) }
+			value={ borderColor || style?.border?.color }
+			disableCustomColors={ disableCustomColors }
+			disableCustomGradients={ disableCustomGradients }
+			onColorChange={ onChangeColor }
+		/>
+	);
+}
+
+/**
+ * Determines if there is border color support.
+ *
+ * @param  {string|Object} blockType Block name or Block Type object.
+ * @return {boolean}                 Whether there is support.
+ */
+export function hasBorderColorSupport( blockType ) {
+	const support = getBlockSupport( blockType, BORDER_SUPPORT_KEY );
+	return !! ( true === support || support?.color );
+}
+
+/**
+ * Custom hook that checks if border color settings have been disabled.
+ *
+ * @param  {string} name The name of the block.
+ * @return {boolean}     Whether border color setting is disabled.
+ */
+export function useIsBorderColorDisabled( { name: blockName } = {} ) {
+	const isDisabled = ! useEditorFeature( 'border.customColor' );
+	return ! hasBorderColorSupport( blockName ) || isDisabled;
+}
+
+/**
+ * Filters registered block settings, extending attributes to include
+ * `borderColor` if needed.
+ *
+ * @param  {Object} settings Original block settings.
+ * @return {Object}          Updated block settings.
+ */
+function addAttributes( settings ) {
+	if ( ! hasBlockSupport( settings, BORDER_COLOR_SUPPORT_KEY ) ) {
+		return settings;
+	}
+
+	// Allow blocks to specify default value if needed.
+	if ( ! settings.attributes.borderColor ) {
+		Object.assign( settings.attributes, {
+			borderColor: {
+				type: 'string',
+			},
+		} );
+	}
+
+	return settings;
+}
+
+/**
+ * Override props assigned to save component to inject border color.
+ *
+ * @param  {Object} props      Additional props applied to save element.
+ * @param  {Object} blockType  Block type definition.
+ * @param  {Object} attributes Block's attributes
+ * @return {Object}            Filtered props to apply to save element.
+ */
+function addSaveProps( props, blockType, attributes ) {
+	if ( ! hasBlockSupport( blockType, BORDER_COLOR_SUPPORT_KEY ) ) {
+		return props;
+	}
+
+	const { borderColor, style } = attributes;
+	const borderColorClass = getColorClassName( 'border-color', borderColor );
+
+	const newClassName = classnames( props.className, {
+		'has-border-color': borderColor || style?.border?.color,
+		[ borderColorClass ]: !! borderColorClass,
+	} );
+
+	props.className = newClassName ? newClassName : undefined;
```

Comment:
what is the intention behind this check?

Reply:
It is to prevent an empty string being assigned to `className` when the named border colors are being toggled on and off while no other classes are present on the block.

Reply:
Wouldn't it be more straightforward to to write it:
```js
if(newClassName)
   props.className = newClassName;
```
?

Reply:
Unfortunately not. It would fail to clear the previous border color classes from the `className` attribute when they were the only ones present.

Take a situation like:
1. Block only has border color classes in its `className`
2. User clears border color selection
3. The new calculated class name in `newClassName` is correctly empty.
4. In the above snippet, it would be falsey and so not clear the `className` prop as required.

I've added another comment to the code to explain this a little more. Hopefully that helps some.